typedb/connection/cluster/stub.py

```diff
@@ -18,51 +18,84 @@
+        self._token = None
+        try:
+            self._token = self._cluster_stub.user_token_renew(self._credential.username())
+        except RpcError as e:
+            e2 = TypeDBClientException.of_rpc(e)
+            if e2.error_message is not None and e2.error_message is not UNABLE_TO_CONNECT:
+                raise e2
 
     def servers_all(self, req: cluster_server_proto.ServerManager.All.Req) -> cluster_server_proto.ServerManager.All.Res:
-        return resilient_call(lambda: self._clusterStub.servers_all(req))
+        return self.resilient_call(lambda: self._cluster_stub.servers_all(req))
 
     def databases_get(self, req: cluster_database_proto.ClusterDatabaseManager.Get.Req) -> cluster_database_proto.ClusterDatabaseManager.Get.Res:
-        return resilient_call(lambda: self._clusterStub.databases_get(req))
+        return self.resilient_call(lambda: self._cluster_stub.databases_get(req))
 
     def databases_all(self, req: cluster_database_proto.ClusterDatabaseManager.All.Req) -> cluster_database_proto.ClusterDatabaseManager.All.Res:
-        return resilient_call(lambda: self._clusterStub.databases_all(req))
+        return self.resilient_call(lambda: self._cluster_stub.databases_all(req))
 
     def users_all(self, req: cluster_user_proto.ClusterUserManager.All.Req) -> cluster_user_proto.ClusterUserManager.All.Res:
-        return resilient_call(lambda: self._clusterStub.users_all(req))
+        return self.resilient_call(lambda: self._cluster_stub.users_all(req))
 
     def users_contains(self, req: cluster_user_proto.ClusterUserManager.Contains.Req) -> cluster_user_proto.ClusterUserManager.Contains.Res:
-        return resilient_call(lambda: self._clusterStub.users_contains(req))
+        return self.resilient_call(lambda: self._cluster_stub.users_contains(req))
 
     def users_create(self, req: cluster_user_proto.ClusterUserManager.Create.Req) -> cluster_user_proto.ClusterUserManager.Create.Res:
-        return resilient_call(lambda: self._clusterStub.users_create(req))
+        return self.resilient_call(lambda: self._cluster_stub.users_create(req))
 
     def user_password(self, req: cluster_user_proto.ClusterUser.Delete.Req) -> cluster_user_proto.ClusterUser.Delete.Res:
-        return resilient_call(lambda: self._clusterStub.user_password(req))
+        return self.resilient_call(lambda: self._cluster_stub.user_password(req))
 
     def user_delete(self, req: cluster_user_proto.ClusterUser.Delete.Req) -> cluster_user_proto.ClusterUser.Delete.Res:
-        return resilient_call(lambda: self._clusterStub.user_delete(req))
+        return self.resilient_call(lambda: self._cluster_stub.user_delete(req))
+
+    def channel(self) -> Channel:
+        return self._channel
+
+    def stub(self) -> TypeDBStub:
+        return self._stub
+
+    def token(self):
+        return self._token
 
+    def resilient_call(self, function: Callable[[], T]) -> T:
```

Comment:
The logic seems sound, but I'm not convinced that this method is the right place to put that logic.

`resilient_call` is so named because in `client-java`, we ensure that the connection is re-established if it has dropped. Checking authentication seems a little too far from that.